transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,356 @@
+    cmp::Reverse,
+    collections::BinaryHeap,
+    future::poll_fn,
+    sync::{Arc, Mutex},
+    task::Poll,
+    time::Duration,
+};
+use tracing::trace;
+
+use crate::data::DelayedData;
+
+#[cfg(all(feature = "prometheus", not(test)))]
+use hopr_metrics::metrics::SimpleGauge;
+
+#[cfg(all(feature = "prometheus", not(test)))]
+lazy_static::lazy_static! {
+    static ref METRIC_QUEUE_SIZE: SimpleGauge =
+        SimpleGauge::new("hopr_mixer_queue_size", "Current mixer queue size").unwrap();
+    static ref METRIC_MIXER_AVERAGE_DELAY: SimpleGauge = SimpleGauge::new(
+        "hopr_mixer_average_packet_delay",
+        "Average mixer packet delay averaged over a packet window"
+    )
+    .unwrap();
+}
+
+pub const METRIC_DELAY_WINDOW: usize = 100;
+pub const HOPR_MIXER_CAPACITY: usize = 10000;
+
+/// Mixing and delaying channel using random delay function.
+///
+/// Mixing is performed by assigning random delays to the ingress timestamp of data,
+/// then storing the values inside a binary heap with reversed ordering (max heap).
+/// This effectively creates a min heap behavior, which is required to ensure that
+/// data is released in order of their delay expiration.
+///
+/// When data arrives:
+/// 1. A random delay is assigned
+/// 2. Data is stored in the heap with its release timestamp
+/// 3. The heap maintains ordering so items with earliest release time are at the top
+///
+/// The channel uses a single timer thread that is instantiated on the first
+/// timer reset and shared across all operations.
+struct Channel<T> {
+    /// Buffer holding the data with a timestamp ordering to ensure the min heap behavior.
+    buffer: BinaryHeap<Reverse<DelayedData<T>>>,
+    timer: futures_timer::Delay,
+    waker: Option<std::task::Waker>,
+    sender_count: usize,
+    receiver_active: bool,
+}
+
+/// Error returned by the [`Sender`].
+#[derive(Debug, thiserror::Error)]
+pub enum SenderError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Sender object interacting with the mixing channel.
+pub struct Sender<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        let mut channel = self.channel.lock().unwrap();
+        channel.sender_count += 1;
+
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        let mut channel = self.channel.lock().unwrap();
+        channel.sender_count -= 1;
+        if channel.sender_count == 0 && !channel.receiver_active {
+            channel.waker = None;
+        }
+    }
+}
+
+impl<T> Sender<T> {
+    /// Send one item to the mixing channel.
+    pub fn send(&self, item: T) -> Result<(), SenderError> {
+        let mut sender = self.clone();
+        sender.start_send_unpin(item)
+    }
```

Comment:
This is very tricky to use in tests. In production, items are most likely getting sent via `SinkExt::send` (or the like) and that has very different semantics - not just by using `poll_ready` first, but actually it also buffers elements in specific way.

Please try to write some additional tests that do not use this `Sender::send`, but use `SinkExt::send`.

Reply:
I will, good idea.